lib/eventHubSender.ts

```diff
@@ -0,0 +1,335 @@
+      return Promise.reject(err);
+    }
+  }
+
+  /**
+   * "Unlink" this sender, closing the link and resolving when that operation is complete.
+   * Leaves the underlying connection/session open.
+   * @method close
+   * @return {Promise<void>} Promise<void>
+   */
+  async close(): Promise<void> {
+    if (this._sender) {
+      try {
+        await rheaPromise.closeSender(this._sender);
+        this.removeAllListeners();
+        delete this._context.senders[this.name!];
+        debug(`Deleted the sender "${this.name!}" from the client cache.`);
+        this._sender = undefined;
+        this._session = undefined;
+        clearTimeout(this._tokenRenewalTimer as NodeJS.Timer);
+        debug(`[${this._context.connectionId}] Sender "${this.name}" closed.`);
+      } catch (err) {
+        return Promise.reject(err);
+      }
+    }
+  }
+
+  /**
+   * Tries to send the message to EventHub if there is enough credit to send them
+   * and the circular buffer has available space to settle the message after sending them.
+   *
+   * We have implemented a synchronous send over here. We shall be waiting for the message
+   * to be accepted or rejected and accordingly resolve or reject the promise.
+   *
+   * @param message The message to be sent to EventHub.
+   * @return {Promise<rheaPromise.Delivery>} Promise<rheaPromise.Delivery>
+   */
+  private _trySend(message: AmqpMessage, tag?: any, format?: number): Promise<rheaPromise.Delivery> {
+    return new Promise((resolve, reject) => {
+      debug(`[${this._context.connectionId}] Sender "${this.name}", credit: ${this._sender.credit}, ` +
+        `available: ${this._sender.session.outgoing.available()}.`);
+      if (this._sender.sendable()) {
+        debug(`[${this._context.connectionId}] Sender "${this.name}", sending message: \n`, message);
+        let onRejected: Func<rheaPromise.Context, void>;
+        const onAccepted: Func<rheaPromise.Context, void> = (context: rheaPromise.Context) => {
+          // Since we will be adding listener for accepted and rejected event every time
+          // we send a message, we need to remove listener for both the events.
+          // This will ensure duplicate listeners are not added for the same event.
+          this._sender.removeListener("accepted", onAccepted);
+
+          this._sender.removeListener("rejected", onRejected);
+          debug(`[${this._context.connectionId}] Sender "${this.name}", got event accepted.`);
+          resolve(context.delivery);
+        };
+        onRejected = (context: rheaPromise.Context) => {
+          this._sender.removeListener("rejected", onRejected);
+          this._sender.removeListener("accepted", onAccepted);
+          debug(`[${this._context.connectionId}] Sender "${this.name}", got event accepted.`);
+          reject(errors.translate(context!.delivery!.remote_state.error));
+        };
+        this._sender.on("accepted", onAccepted);
+        this._sender.on("rejected", onRejected);
```

Comment:
there are actually [4 dispositions in amqp](http://www.amqp.org/sites/amqp.org/files/amqp.pdf?page=82) (accepted, rejected, modified, released). does rhea expose events for all four?

Reply:
reject Promise on modified and released. If error is present then bubble it up otherwise wrap it by letting the customer know that it was modified/released and not accepted.

Reply:
5. received is a non-terminal state for messages spanning accross multiple frames

---
In reply to: [180829472](https://github.com/Azure/azure-event-hubs-node/pull/47#discussion_r180829472) [](ancestors = 180829472)